Gems/PhysX/Code/Editor/Source/ComponentModes/Joints/JointsComponentMode.cpp

```diff
@@ -0,0 +1,598 @@
+                        [this]()
+                        {
+                            SetCurrentMode(
+                                JointsComponentModeCommon::SubComponentModes::ModeType::SnapRotation,
+                                m_buttonData[JointsComponentModeCommon::SubComponentModes::ModeType::SnapRotation]);
+                        }));
+                }
+                break;
+            }
+        }
+
+        // reset values
+        {
+            AzToolsFramework::ActionOverride resetValuesAction = makeActionOverride(
+                entityComponentIdPair, SubModeData::ResetSubMode,
+                SubModeData::ResetTitle, SubModeData::ResetToolTip,
+                [this]()
+                {
+                    ResetCurrentMode();
+                });
+            resetValuesAction.SetKeySequence(QKeySequence(Qt::Key_R));
+            actions.emplace_back(resetValuesAction);
+        }
+
+        return actions;
+    }
+
+    AZStd::vector<AzToolsFramework::ViewportUi::ClusterId> JointsComponentMode::PopulateViewportUiImpl()
+    {
+        return m_modeSelectionClusterIds;
+    }
+
+    void JointsComponentMode::SetCurrentMode(JointsComponentModeCommon::SubComponentModes::ModeType newMode, ButtonData& buttonData)
+    {
+        auto subMode = m_subModes.find(newMode);
+        if (subMode != m_subModes.end())
+        {
+            const AZ::EntityComponentIdPair entityComponentIdPair = GetEntityComponentIdPair();
+            m_subModes[m_subMode]->Teardown(entityComponentIdPair);
+            m_subMode = newMode;
+            subMode->second->Setup(entityComponentIdPair);
+
+            // if this button is on a different cluster. clear the active state.
+            if (m_activeButton.m_clusterId != buttonData.m_clusterId)
+            {
+                AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+                    AzToolsFramework::ViewportUi::DefaultViewportId,
+                    &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::ClearClusterActiveButton, m_activeButton.m_clusterId);
+            }
+            AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+                AzToolsFramework::ViewportUi::DefaultViewportId,
+                &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterActiveButton, buttonData.m_clusterId,
+                buttonData.m_buttonId);
+            m_activeButton = buttonData;
+        }
+        AZ_Assert(subMode != m_subModes.end(), "PhysXJoints Uninitialized joint component mode selected.");
+    }
+
+    bool JointsComponentMode::HandleMouseInteraction(const AzToolsFramework::ViewportInteraction::MouseInteractionEvent& mouseInteraction)
+    {
+        // Propagate mouse interaction to sub-component mode.
+        if (m_subModes[m_subMode])
+        {
+            m_subModes[m_subMode]->HandleMouseInteraction(mouseInteraction);
```

Comment:
question: Should this function be returning the result of `HandleMouseInteraction`?

Reply:
the sub modes ```HandleMouseInteraction``` is setup to not consume the input and the function has a void return type. This is similar to the collider component mode setup.

Reply:
Ah okay perfect, thanks for confirming 👍